lib/ansible/executor/play_iterator.py

```diff
@@ -296,28 +297,26 @@ def _get_next_task_from_state(self, state, host):
                 # during play compilation in __init__ above.
                 # If not, below we will determine if we do in fact want to gather
                 # facts or validate arguments for the specified host.
-                state.pending_setup = state.cur_regular_task < len(block.block)
-                if state.pending_setup:
-                    task = block.block[state.cur_regular_task]
-
+                if state.pending_setup is None:
+                    state.pending_setup = True
                     # Gather facts if the default is 'smart' and we have not yet
                     # done it for this host; or if 'explicit' and the play sets
                     # gather_facts to True; or if 'implicit' and the play does
                     # NOT explicitly set gather_facts to False.
-                    gather_facts = bool(state.cur_regular_task == 0)
+                    task = self._blocks[0].block[0]
                     gathering = C.DEFAULT_GATHERING
                     implied = self._play.gather_facts is None or boolean(self._play.gather_facts, strict=False)
-
-                    if gather_facts and not (
+                    if not (
                         (gathering == 'implicit' and implied) or
                         (gathering == 'explicit' and boolean(self._play.gather_facts, strict=False)) or
                         (gathering == 'smart' and implied and not self._variable_manager._facts_gathered_for_host(host.name))
                     ):
                         task = None
-                    elif not gather_facts and not self._play.validate_argspec:
+                elif state.pending_setup:
+                    task = self._blocks[0].block[1]
+                    if not self._play.validate_argspec:
```

Comment:
I wonder if creating a new state (`IteratingStates.ValidatePlayArgspec` or something) would simplify the implementation?